### PR TITLE
perf(channel): performance improvements 

### PR DESCRIPTION
**Pull request by ramey**
### Changes
The following changes are done as part of this PR -
- Remove serde interfaces to use direct functions
- Fix shutdown
- Change `ProduceBulk` method to remove `CollectRequest`

Closing #35

### CODE REVIEW

**Comment by NNcrawler** [comment]
The PR is LGTM. Can you also put the benchmark result as well in the issue?

File: collection/collector.go [hunk 1]
```diff
@@ -5,14 +5,23 @@ import (
...
+	Type       string
+	EventBytes []byte
+}
+
 type CollectRequest struct {
 	ConnectionIdentifier identification.Identifier
 	TimeConsumed         time.Time
 	TimePushed           time.Time
-	*pb.SendEventRequest
+	SentTime             time.Time
```
Comment (chakravarthyvp) [inline collection/collector.go:19]: https://github.com/odpf/raccoon/issues/35#issuecomment-1141682794
Inlining with this comment for reference

**Comment by ramey** [comment]
## Performace Benchmarking

### **Setup**

Max Clients - 1000
Batch Size - 100
Batch Interval - 10000
Execution time - 30m

### v0.1.1

**Server Processing Latency(max_upper)**
Screenshot 2022-06-10 at 12 34 11 PM

Min 624ms
Max - 6.71 min
Avg - 45.8s

**Batch idle time in channel**
Screenshot 2022-06-10 at 12 34 24 PM

Min - 1.03 min
Max - 7.6 min
Avg - 5.6 min

**Worker Latency(max_upper)**
Screenshot 2022-06-10 at 12 47 38 PM

Min 191 ms
Max 1.02 min
Mean 8.92s

### v0.1.3

**Server Processing Latency(max_upper)**
Screenshot 2022-06-10 at 3 06 09 PM

Min - 6.74s
Max - 17.9 min
Avg 7.79 min

**Batch idle time in channel**
Screenshot 2022-06-10 at 3 06 04 PM

Min - 2.20  min
Max - 9.09 min
Avg - 6.54 min

**Worker Latency(max_upper)**
Screenshot 2022-06-10 at 3 05 57 PM

Min - 100ms
Max - 1 min
Avg - 13.2 s

### Current Branch

**Server Processing Latency(max_upper)**
Screenshot 2022-06-12 at 5 28 38 PM

Min - 15ms
Max - 7.04 min
Avg - 1min

**Batch idle time in channel**
Screenshot 2022-06-12 at 5 28 46 PM

Min - 1.63s
Max - 7.08 min
Avg - 3.69 min

**Worker Latency(max_upper)**
Screenshot 2022-06-12 at 5 29 04 PM

Min 58ms
Max 3.04s
Avg 512 ms

File: services/rest/service.go [hunk 2]
```diff
@@ -52,14 +37,35 @@ func (s Service) Init(ctx context.Context) error {
...
+	t := time.Tick(config.MetricStatsd.FlushPeriodMs)
+	for {
+		<-t
+		for k, v := range conn.TotalConnectionPerGroup() {
+			metrics.Gauge("connections_count_current", v, fmt.Sprintf("conn_group=%s", k))
+		}
+	}
+}
+
+func (s *Service) Init(ctx context.Context) error {
```
Comment (siddhanta-rath) [inline services/rest/service.go:61]: `ctx` is unused and can be removed

File: services/rest/service.go [hunk 3]
```diff
@@ -52,14 +37,35 @@ func (s Service) Init(ctx context.Context) error {
...
+	return s.s.ListenAndServe()
 }
 
-func (s Service) Name() string {
+func (s *Service) Name() string {
```
Comment (siddhanta-rath) [inline services/rest/service.go:65]: `s` is unused and can be removed

File: publisher/kafka_test.go [hunk 4]
```diff
@@ -90,7 +89,7 @@ func TestKafka_ProduceBulk(suite *testing.T) {
 			client.On("Produce", mock.Anything, mock.Anything).Return(fmt.Errorf("Local: Unknown topic")).Once()
 			kp := NewKafkaFromClient(client, 10, "%s")
 
-			err := kp.ProduceBulk(collection.CollectRequest{SendEventRequest: &pb.SendEventRequest{Events: []*pb.Event{{EventBytes: []byte{}, Type: topic}}}}, make(chan kafka.Event, 2))
+			err := kp.ProduceBulk([]*pb.Event{{EventBytes: []byte{}, Type: topic}}, "group-1", make(chan kafka.Event, 2))
```
Comment (siddhanta-rath) [inline publisher/kafka_test.go:92]: `group-1` should be declared as constant or global variable as it has been used at many places

File: deserialization/proto.go [hunk 5]
```diff
@@ -8,9 +8,9 @@ import (
 
 var ErrInvalidProtoMessage = errors.New("invalld proto message")
 
-type ProtoDeserilizer struct{}
+// type ProtoDeserilizer struct{}
 
-func (d *ProtoDeserilizer) Deserialize(b []byte, i interface{}) error {
+func DeserializeProto(b []byte, i interface{}) error {
 	msg, ok := i.(proto.Message)
 	if !ok {
```
Comment (siddhanta-rath) [inline deserialization/proto.go:15]: nice 👍

File: collection/collector.go [hunk 6]
```diff
@@ -8,6 +8,11 @@ import (
 	pb "github.com/odpf/raccoon/proto"
 )
 
+type Event struct {
+	Type       string
+	EventBytes []byte
```
Comment (siddhanta-rath) [inline collection/collector.go:13]: will `data ` or `content` be a better name instead of `EventBytes`

File: collection/collector.go [hunk 7]
```diff
@@ -8,6 +8,11 @@ import (
 	pb "github.com/odpf/raccoon/proto"
 )
 
+type Event struct {
```
Comment (chakravarthyvp) [inline collection/collector.go:11]: Wasn't this supposed to be removed?

File: services/pprof/service.go [hunk 8]
```diff
@@ -3,6 +3,7 @@ package pprof
 import (
 	"context"
 	"net/http"
+
 	// enable pprof https://pkg.go.dev/net/http/pprof#pkg-overview
 	_ "net/http/pprof"
```
[the remark this answers is not in the document]
Comment (ramey) [reply services/pprof/service.go:8]: This is fine right, we want to expose the paths by default.

File: publisher/kafka.go [hunk 9]
```diff
@@ -64,26 +63,26 @@ func (pr *Kafka) ProduceBulk(request collection.CollectRequest, deliveryChannel
 
 		err := pr.kp.Produce(message, deliveryChannel)
 		if err != nil {
-			metrics.Increment("kafka_messages_delivered_total", fmt.Sprintf("success=false,conn_group=%s,event_type=%s", request.ConnectionIdentifier.Group, event.GetType()))
+			metrics.Increment("kafka_messages_delivered_total", fmt.Sprintf("success=false,conn_group=%s,event_type=%s", connGroup, event.Type))
 			if err.Error() == "Local: Unknown topic" {
 				errors[order] = fmt.Errorf("%v %s", err, topic)
-				metrics.Increment("kafka_unknown_topic_failure_total", fmt.Sprintf("topic=%s,event_type=%s,conn_group=%s", topic, event.Type, request.ConnectionIdentifier.Group))
+				metrics.Increment("kafka_unknown_topic_failure_total", fmt.Sprintf("topic=%s,event_type=%s,conn_group=%s", topic, event.Type, connGroup))
```
[the remark this answers is not in the document]
Comment (ramey) [reply publisher/kafka.go:69]: Agree with you on this, actually `fmt.Sprintf` is used all over Raccoon and this PR might not be the right one to do the change. Maybe we can create a new Issue and put the details there and then raise a PR for the same.

File: services/grpc/service.go [hunk 10]
```diff
@@ -16,22 +16,28 @@ type Service struct {
...
+func NewGRPCService(c collection.Collector) *Service {
+	server := grpc.NewServer()
+	pb.RegisterEventServiceServer(server, &Handler{C: c})
+	return &Service{
+		s:         server,
+		Collector: c,
+	}
+}
+
+func (s *Service) Init(ctx context.Context) error {
```
[the remark this answers is not in the document]
Comment (ramey) [reply services/grpc/service.go:28]: This is an implementation of the `bootstraper` interface,  which has `Init(ctx context.Context) error` as the function signature

File: services/pprof/service.go [hunk 11]
```diff
@@ -11,10 +12,14 @@ type Service struct {
 	s *http.Server
 }
 
+func NewPprofService() Service {
+	return Service{
+		s: &http.Server{Addr: "localhost:6060", Handler: nil},
+	}
+}
+
 func (s Service) Init(ctx context.Context) error {
```
[the remark this answers is not in the document]
Comment (ramey) [reply services/pprof/service.go:21]: Changing it to pointer

File: services/rest/service.go (hunk 3 above)
Comment (ramey) [reply services/rest/service.go:65]: Same, this is an interface implementation.

File: publisher/kafka_test.go (hunk 4 above)
Comment (ramey) [reply publisher/kafka_test.go:92]: changed

File: collection/collector.go (hunk 6 above)
Comment (ramey) [reply collection/collector.go:13]: This interface is stale, so will remove this. Anyway this came from [here](https://github.com/odpf/proton/blob/main/odpf/raccoon/v1beta1/raccoon.proto#L68)

**Pull request merged**